include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,502 @@
+ * @param id Id of the alarm
+ * @param mask Mask of fields in the alarm time which are supported
+ *
+ * @note Bits in the mask param are defined here \ref rtc_alarm_time_mask
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range or time is invalid
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_get_supported_fields(const struct device *dev, uint16_t id,
+					     uint16_t *mask);
+
+static inline int z_impl_rtc_alarm_get_supported_fields(const struct device *dev, uint16_t id,
+							uint16_t *mask)
+{
+	const struct rtc_driver_api *api = (const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_get_supported_fields(dev, id, mask);
+}
+
+/**
+ * @brief API for setting RTC alarm time.
+ *
+ * @details To enable an RTC alarm, one or more fields of the RTC alarm time
+ * must be enabled. The mask designates which fields of the RTC alarm time to
+ * enable. If the mask parameter is 0, the alarm will be disabled. The RTC
+ * alarm will trigger when all enabled fields of the alarm time match the RTC
+ * time.
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm
+ * @param mask Mask of fields in the alarm time to enable
+ * @param timeptr The alarm time to set
+ *
+ * @note The timeptr param may be NULL if the mask param is 0
+ * @note Only the enabled fields in the timeptr param need to be configured
+ * @note Bits in the mask param are defined here \ref rtc_alarm_time_mask
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range or time is invalid
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+				 const struct rtc_time *timeptr);
+
+static inline int z_impl_rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+					    const struct rtc_time *timeptr)
+{
+	const struct rtc_driver_api *api = (const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_set_time(dev, id, mask, timeptr);
+}
+
+/**
+ * @brief API for getting RTC alarm time.
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm
+ * @param mask Destination for mask of fields which are enabled in the alarm time
+ * @param timeptr Destination for the alarm time
+ *
+ * @note Bits in the mask param are defined here \ref rtc_alarm_time_mask
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				 struct rtc_time *timeptr);
+
+static inline int z_impl_rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+					    struct rtc_time *timeptr)
+{
+	const struct rtc_driver_api *api = (const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_get_time(dev, id, mask, timeptr);
```

Comment:
No, this needs to be implemented properly, it needs to check if the function exists and only call it if it does, otherwise return -NOSYS. If I have 2 RTCs, one with alarms and one without, calling this on the wrong one is completely undefined operation. This needs fixing for all the optional functions, the function pointers themselves do not need to be controlled by if a Kconfig is enabled, these functions need the protection code in

Reply:
According to the documentation https://docs.zephyrproject.org/3.0.0/reference/api/design_guidelines.html (Return codes section), drivers must implement all API functions, and return -ENOSYS or -ENOTSUP. According to the documentation, to my knowledge, it is not up to the API wrapper to protect against the driver not implementing unsupported API endpoints.

However, some driver APIs do the check anyway, like the bbram and gpio, while others don't, like I2C, sensor and SPI. Which is the proper implementation is seemingly at odds between some drivers and the documentation.

I will add it in this case, with the argument that there is precedence for it, and that the RTC is usually not accessed at a high enough frequency to cause a noticeable performance impact, compared to something like SPI and I2C.

Reply:
I think there's a difference between "optional" API methods and mandatory ones.  There is no check for methods all drivers must implement, but if it's common for drivers to not implement a method, then the wrapper should check and return ENOSYS.  I already described this and there are quite a few APIs that do this.

IMHO, alarm related methods, check.  Set and get time, no check.

Reply:
All functions except `set_time` and `get_time` have been updated to validate API has indeed been implemented.